lxd/instance/drivers/driver_lxc.go

```diff
@@ -1829,6 +1874,95 @@ func (d *lxc) DeviceEventHandler(runConf *deviceConfig.RunConfig) error {
 	return nil
 }
 
+func (d *lxc) handleIdmappedStorage() (idmap.IdmapStorageType, *idmap.IdmapSet, error) {
+	diskIdmap, err := d.DiskIdmap()
+	if err != nil {
+		return idmap.IdmapStorageNone, nil, errors.Wrap(err, "Set last ID map")
+	}
+
+	nextIdmap, err := d.NextIdmap()
+	if err != nil {
+		return idmap.IdmapStorageNone, nil, errors.Wrap(err, "Set ID map")
+	}
+
+	// privileged container
+	if diskIdmap == nil && nextIdmap == nil {
+		return idmap.IdmapStorageNone, nil, nil
+	}
+
+	idmapType := d.IdmappedStorage(d.RootfsPath())
+
+	// Unprivileged container with identical idmapping and the kernel
+	// doesn't support idmapped storage.
+	if diskIdmap != nil && nextIdmap.Equals(diskIdmap) && idmapType == idmap.IdmapStorageNone {
+		return idmap.IdmapStorageNone, nil, nil
+	}
```

Comment:
So this means we will change an on-disk idmapped container to an idmapped storage container unless `security.protection.shift` is set.

Reply:
Which is how I understood our previous logic did it just in a very unreadable way.